src/CallbackContainer.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+namespace Dhii\Container;
+
+use Dhii\Container\Util\StringTranslatingTrait;
+use Psr\Container\ContainerInterface;
+use function call_user_func_array;
+
+/**
+ * A container implementation that decorates another container, invoking a callback for possible modifications to the
+ * decorated container's values.
+ *
+ * @since [*next-version*]
+ */
+class CallbackContainer implements ContainerInterface
+{
+    /* @since [*next-version*] */
+    use StringTranslatingTrait;
+
+    /**
+     * @since [*next-version*]
+     *
+     * @var callable
+     */
+    protected $callback;
+
+    /**
+     * @since [*next-version*]
+     *
+     * @var ContainerInterface
+     */
+    protected $inner;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param ContainerInterface|null $inner      The container instance to decorate.
+     * @param callable                $callback   The callback to invoke on get. It will be passed 3 parameters:
+     *                                            * The inner container's value for the key being fetched.
+     *                                            * The key being fetched.
+     *                                            * A reference to this container instance.
+     */
+    public function __construct(ContainerInterface $inner, callable $callback)
+    {
+        $this->callback = $callback;
+        $this->inner = $inner;
+    }
+
+    /**
+     * @inheritdoc
+     *
+     * @since [*next-version*]
+     */
+    public function get($key)
+    {
+        return call_user_func_array($this->callback, [$this->inner->get($key), $key, $this]);
```

Comment:
I'm thinking to start using `$this->callback($this->inner->get($key), $key, $this)`. AFAIK it's functionally identical, but simpler. Thoughts?

Reply:
I'd prefer the simpler syntax too. Need to be careful about it though as some linters or code sniffers might complain about it, since the syntax is identical to calling a method named `callback`.

Reply:
That's true. So, maybe store it temporarily, or try the curly braces or just parens?

Reply:
I found an even better way: Parentheses!

```php
return ($this->callback)(...);
```